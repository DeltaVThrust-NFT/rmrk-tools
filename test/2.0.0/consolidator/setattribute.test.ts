import { Consolidator, NFT } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  getAliceKey,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock,
  mintNftMock2,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";

beforeAll(async () => {
  return await cryptoWaitReady();
});

const mintNftWithPropertiesIssuer = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutator: "issuer",
        value: "foo",
        type: "string",
      },
    },
  });

const mintNftWithPropertiesOwner = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutator: "owner",
        value: "foo",
        type: "string",
      },
    },
  });

const mintNftWithProperties = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        value: "foo",
        type: "string",
      },
    },
  });

const mintNftWithProperties2 = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutator: "boo",
        value: "foo",
        type: "string",
      },
    },
  });

describe("rmrk2.0.0 Consolidator: SETATTRIBUTE", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
  ];

  it("Mint NFT with properties", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesIssuer().mint()),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should throw if we are trying to mutate immutable attribute", async () => {
    expect(() =>
      mintNftWithProperties(3).setattribute("test", {
        type: "int",
        value: 2,
      })
    ).toThrow();

    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithProperties2().mint()),
      ...getBlockCallsMock(
        mintNftWithProperties2(3).setattribute("test", {
          type: "int",
          value: 2,
        })
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETATTRIBUTE] Attempting to set attribute on immutable attribute test"
    );
  });

  it("should throw if we are trying to mutate attribute with wrong rootowner _mutator", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(3).setattribute("test", {
          type: "int",
          value: 2,
        }),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETATTRIBUTE] Attempting to set attribute on and NFT where rootowner is _mutator don't match. Expected 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty but received 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    );
  });

  it("should throw if we are trying to mutate attribute with wrong issuer _mutator", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesIssuer().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesIssuer(3).setattribute("test", {
          type: "int",
          value: 2,
        }),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETATTRIBUTE] Attempting to set attribute on and NFT where issuer is _mutator don't match. Expected 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty but received 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    );
  });

  it("should allow to mutate attribute if issuer _mutator match", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesIssuer().mint()),
      ...getBlockCallsMock(mintNftMock2(3).send(getBobKey().address)), // Send to Bob first
      ...getBlockCallsMock(
        mintNftWithPropertiesIssuer(3).setattribute("test", {
          type: "int",
          value: 2,
        })
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should allow to mutate attribute if owner _mutator match", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(mintNftMock2(3).send(getBobKey().address)), // Send to Bob first
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(3).setattribute("test", {
          type: "int",
          value: 2,
        }),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });
});
