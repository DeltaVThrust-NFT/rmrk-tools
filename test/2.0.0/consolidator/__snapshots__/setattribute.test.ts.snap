// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: SETATTRIBUTE Mint NFT with properties 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutator": "issuer",
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: SETATTRIBUTE should allow to mutate attribute if issuer _mutator match 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 4,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[SEND] Attempting to send non-existant NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "op_type": "SEND",
    },
  ],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "type": "int",
          "value": 2,
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: SETATTRIBUTE should allow to mutate attribute if owner _mutator match 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [
    Object {
      "block": 4,
      "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "message": "[SEND] Attempting to send non-existant NFT 3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "op_type": "SEND",
    },
    Object {
      "block": 5,
      "caller": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "message": "[SETATTRIBUTE] Attempting to set attribute on and NFT where rootowner is _mutator don't match. Expected 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty but received 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "object_id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "op_type": "SETATTRIBUTE",
    },
  ],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutator": "owner",
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;
